translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,388 @@
+				attrs := ocAttrsToInternal(teValue.Annotation.Attributes)
+				event.SetAttributes(attrs)
+			}
+
+		case *octrace.Span_TimeEvent_MessageEvent_:
+			event.SetAttributes(ocMessageEventToInternalAttrs(teValue.MessageEvent))
+
+		default:
+			event.SetName("An unknown OpenCensus TimeEvent type was detected when translating")
+		}
+	}
+	return
+}
+
+func ocLinksToInternal(ocLinks *octrace.Span_Links) (ptrs []*data.SpanLink, droppedCount uint32) {
+	if ocLinks == nil {
+		return
+	}
+
+	droppedCount = uint32(ocLinks.DroppedLinksCount)
+
+	linkCount := len(ocLinks.Link)
+	if linkCount == 0 {
+		return
+	}
+
+	// Create slice that holds pointers and another slice that holds structs at once
+	// to avoid individual allocations of structs.
+	ptrs = make([]*data.SpanLink, 0, linkCount)
+	content := data.NewSpanLinkSlice(linkCount)
+
+	for i, ocLink := range ocLinks.Link {
+		if ocLink == nil {
+			continue
+		}
+
+		// Point one element in slice of pointers to an element in slice of structs.
+		link := &content[i]
+		ptrs = append(ptrs, link)
+
+		attrs := ocAttrsToInternal(ocLink.Attributes)
+		link.SetTraceID(data.TraceIDFromBytes(ocLink.TraceId))
+		link.SetSpanID(data.SpanIDFromBytes(ocLink.SpanId))
+		link.SetTraceState(ocTraceStateToInternal(ocLink.Tracestate))
+		link.SetAttributes(attrs)
+	}
+	return
+}
+
+func ocMessageEventToInternalAttrs(msgEvent *octrace.Span_TimeEvent_MessageEvent) data.Attributes {
+	if msgEvent == nil {
+		return data.Attributes{}
+	}
+
+	attrs := data.NewAttributeValueSlice(4)
+	attrs[0].SetString(msgEvent.Type.String())
+	attrs[1].SetInt(int64(msgEvent.Id))
+	attrs[2].SetInt(int64(msgEvent.UncompressedSize))
+	attrs[3].SetInt(int64(msgEvent.CompressedSize))
+
+	return data.NewAttributes(
+		map[string]data.AttributeValue{
+			conventions.OcTimeEventMessageEventType:  attrs[0],
+			conventions.OcTimeEventMessageEventID:    attrs[1],
+			conventions.OcTimeEventMessageEventUSize: attrs[2],
+			conventions.OcTimeEventMessageEventCSize: attrs[3],
+		},
+		0)
+}
+
+func truncableStringToStr(ts *octrace.TruncatableString) string {
+	if ts == nil {
+		return ""
+	}
+	return ts.Value
+}
+
+func ocNodeResourceToInternal(ocNode *occommon.Node, ocResource *ocresource.Resource) *data.Resource {
+	resource := &data.Resource{}
+
+	// Number of special fields in the Node. See the code below that deals with special fields.
+	const specialNodeAttrCount = 7
+
+	// Number of special fields in the Resource.
+	const specialResourceAttrCount = 1
+
+	// Calculate maximum total number of attributes. It is OK if we are a bit higher than
+	// the exact number since this is only needed for capacity reservation.
+	maxTotalAttrCount := 0
+	if ocNode != nil {
+		maxTotalAttrCount += len(ocNode.Attributes) + specialNodeAttrCount
+	}
+	if ocResource != nil {
+		maxTotalAttrCount += len(ocResource.Labels) + specialResourceAttrCount
+	}
+
+	// Create a map where we will place all attributes from the Node and Resource.
```

Comment:
This is just a hint to the map, I don't think bailing out is important here.

Reply:
basically it is removing the `if len(attrs0) != 0 {` that you have below by a check here against ``, it is a nit anyway up to you.

Reply:
@pjanotti You could probably just add a check to the top:
```
if ocNode == nil && ocResource == nil {
  return nil
}
```
Then you won't need to check `if len(attrs) != 0` below indeed.
This code is much easier to follow than checking `maxTotalAttrCount` and trying to explicitly avoid allocation here IMO.

Reply:
@nilebox my idea was to use the Get functions so it would cover both when the instances were nil and the maps were empty, that said I have to recognize that my idea seems convoluted :) since it wasn't clear to you or @tigrannajaryan - perhaps because it is connected to the previous suggestion.

Reply:
I think the code as is now is straightforward. I'd prefer to keep it as is.